Neos.EventSourcedContentRepository/Classes/Domain/ValueObject/PropertyValue.php

```diff
@@ -48,7 +48,22 @@ public static function fromArray(array $valueAndType): self
         if (!array_key_exists('type', $valueAndType)) {
             throw new \InvalidArgumentException('Missing array key "type"', 1546524609);
         }
-        return new static($valueAndType['value'], $valueAndType['type']);
+
+        $value = self::convertValueToObject($valueAndType['value'], $valueAndType['type']);
+
+        return new static($value, $valueAndType['type']);
+    }
+
+    private static function convertValueToObject($value, $type)
+    {
+        if ($type === 'DateTime' && is_array($value) && isset($value['date']) && isset($value['timezone']) && isset($value['dateFormat'])) {
+            return \DateTime::createFromFormat($value['dateFormat'], $value['date'], new \DateTimeZone($value['timezone']));
+        }
+
+        if (($type === 'DateTimeImmutable' || $type === 'DateTimeInterface') && is_array($value) && isset($value['date']) && isset($value['timezone']) && isset($value['dateFormat'])) {
+            return \DateTimeImmutable::createFromFormat($value['dateFormat'], $value['date'], new \DateTimeZone($value['timezone']));
+        }
+        return $value;
```

Comment:
As discussed: This should be moved up to the `PropertyCollection`

Side node: I would suggest to change the naming:
`PropertyValue(s)` => `RawPropertyValue(s)`

`PropertyCollection` => `DeserializedPropertyValues` (unfortunately there's a published `PropertyCollectionInterface` so this can't be changed lightly)